File: Colage/Core/University/UniversityService.swift
import Foundation
import SwiftUI

/// Manages university server data and branding
class UniversityService: ObservableObject {
    @Published var currentUniversity: University?
    @Published var currentTheme: UniversityTheme?
    @Published var availableThemes: [UniversityTheme] = []

    /// Resolve university from .edu domain
    func resolveUniversity(domain: String) async -> University? {
        if AppState.devMode {
            return mockUniversity(for: domain)
        }

        // Fetch from API (auto-creates if unknown)
        do {
            struct UniResponse: Decodable { let university: University }
            let result: UniResponse = try await APIClient.shared.request(
                path: "/universities/\(domain)"
            )
            return result.university
        } catch {
            print("Failed to resolve university: \(error)")
            print("⚠️ UniversityService: API failed, using mock data")
            return mockUniversity(for: domain)
        }
    }

    func setUniversity(_ university: University) {
        currentUniversity = university
        availableThemes = university.brandingThemes
        currentTheme = university.brandingThemes.first
    }

    func selectTheme(_ theme: UniversityTheme) {
        currentTheme = theme
    }

    // MARK: - Mock Data (Dev Mode)

    private func mockUniversity(for domain: String) -> University {
        switch domain {
        case "umich.edu":
            return University(
                id: "umich",
                domain: "umich.edu",
                name: "University of Michigan",
                memberCount: 847,
                brandingThemes: [
                    UniversityTheme(
                        id: "umich-maize-blue",
                        name: "Maize & Blue",
                        primaryColor: "#FFCB05",
                        accentColor: "#00274C",
                        textColor: "#FFFFFF",
                        backgroundAsset: nil
                    ),
                    UniversityTheme(
                        id: "umich-stadium",
                        name: "The Big House",
                        primaryColor: "#FFCB05",
                        accentColor: "#00274C",
                        textColor: "#FFFFFF",
                        backgroundAsset: nil
                    ),
                    UniversityTheme(
                        id: "umich-block-m",
                        name: "Block M Classic",
                        primaryColor: "#FFCB05",
                        accentColor: "#00274C",
                        textColor: "#FFFFFF",
                        backgroundAsset: nil
                    )
                ]
            )
        case "harvard.edu":
            return University(
                id: "harvard",
                domain: "harvard.edu",
                name: "Harvard University",
                memberCount: 512,
                brandingThemes: [
                    UniversityTheme(
                        id: "harvard-crimson",
                        name: "Crimson",
                        primaryColor: "#A51C30",
                        accentColor: "#F5F0E1",
                        textColor: "#FFFFFF",
                        backgroundAsset: nil
                    )
                ]
            )
        case "stanford.edu":
            return University(
                id: "stanford",
                domain: "stanford.edu",
                name: "Stanford University",
                memberCount: 623,
                brandingThemes: [
                    UniversityTheme(
                        id: "stanford-cardinal",
                        name: "Cardinal",
                        primaryColor: "#8C1515",
                        accentColor: "#D2C295",
                        textColor: "#FFFFFF",
                        backgroundAsset: nil
                    )
                ]
            )
        default:
            // Auto-create university from domain — new school, first student!
            let shortName = domain
                .replacingOccurrences(of: ".edu", with: "")
                .uppercased()
            return University(
                id: domain,
                domain: domain,
                name: shortName,
                memberCount: 0,
                brandingThemes: [UniversityTheme.default]
            )
        }
    }
}

File: android/app/src/main/java/com/colageclub/colage/core/university/UniversityService.kt
package com.colageclub.colage.core.university

import androidx.compose.ui.graphics.Color
import com.colageclub.colage.BuildConfig
import com.colageclub.colage.core.networking.ApiClient
import com.colageclub.colage.data.models.University
import com.colageclub.colage.data.models.UniversityTheme
import android.util.Log
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class UniversityService @Inject constructor(
    private val apiClient: ApiClient
) {
    private val _currentUniversity = MutableStateFlow<University?>(null)
    val currentUniversity: StateFlow<University?> = _currentUniversity.asStateFlow()

    private val _currentTheme = MutableStateFlow(UniversityTheme.default)
    val currentTheme: StateFlow<UniversityTheme> = _currentTheme.asStateFlow()

    private val _availableThemes = MutableStateFlow<List<UniversityTheme>>(emptyList())
    val availableThemes: StateFlow<List<UniversityTheme>> = _availableThemes.asStateFlow()

    suspend fun resolveUniversity(domain: String): University {
        if (BuildConfig.DEV_MODE) {
            return mockUniversity(domain).also { setUniversity(it) }
        }
        return try {
            val response = apiClient.getUniversity(domain)
            val uni = University(
                id = response.id,
                domain = response.domain,
                name = response.name,
                memberCount = response.memberCount
            )
            setUniversity(uni)
            uni
        } catch (_: Exception) {
            Log.w("UniversityService", "API failed, using mock data")
            mockUniversity(domain).also { setUniversity(it) }
        }
    }

    fun setUniversity(university: University) {
        _currentUniversity.value = university
        val themes = university.brandingThemes.ifEmpty { listOf(UniversityTheme.default) }
        _availableThemes.value = themes
        _currentTheme.value = themes.first()
    }

    fun selectTheme(theme: UniversityTheme) {
        _currentTheme.value = theme
    }

    // MARK: - Mock Data (Dev Mode)

    private fun mockUniversity(domain: String): University {
        return when (domain) {
            "umich.edu" -> University(
                id = "umich",
                domain = "umich.edu",
                name = "University of Michigan",
                memberCount = 847,
                brandingThemes = listOf(
                    UniversityTheme(
                        id = "umich-maize-blue",
                        name = "Maize & Blue",
                        primaryColor = "#FFCB05",
                        accentColor = "#00274C",
                        textColor = "#FFFFFF"
                    ),
                    UniversityTheme(
                        id = "umich-block-m",
                        name = "Block M Classic",
                        primaryColor = "#FFCB05",
                        accentColor = "#00274C",
                        textColor = "#FFFFFF"
                    )
                )
            )
            "harvard.edu" -> University(
                id = "harvard",
                domain = "harvard.edu",
                name = "Harvard University",
                memberCount = 512,
                brandingThemes = listOf(
                    UniversityTheme(
                        id = "harvard-crimson",
                        name = "Crimson",
                        primaryColor = "#A51C30",
                        accentColor = "#F5F0E1",
                        textColor = "#FFFFFF"
                    )
                )
            )
            "stanford.edu" -> University(
                id = "stanford",
                domain = "stanford.edu",
                name = "Stanford University",
                memberCount = 623,
                brandingThemes = listOf(
                    UniversityTheme(
                        id = "stanford-cardinal",
                        name = "Cardinal",
                        primaryColor = "#8C1515",
                        accentColor = "#D2C295",
                        textColor = "#FFFFFF"
                    )
                )
            )
            else -> {
                val shortName = domain.removeSuffix(".edu").uppercase()
                University(
                    id = domain,
                    domain = domain,
                    name = shortName,
                    memberCount = 0,
                    brandingThemes = listOf(UniversityTheme.default)
                )
            }
        }
    }
}

// Extension to parse hex color string to Compose Color
fun UniversityTheme.primaryComposeColor(): Color {
    return try {
        val hex = primaryColor.trimStart('#')
        Color(android.graphics.Color.parseColor("#$hex"))
    } catch (_: Exception) {
        Color(0xFFA51C30)
    }
}

fun UniversityTheme.accentComposeColor(): Color {
    return try {
        val hex = accentColor.trimStart('#')
        Color(android.graphics.Color.parseColor("#$hex"))
    } catch (_: Exception) {
        Color(0xFF00CEC9)
    }
}
